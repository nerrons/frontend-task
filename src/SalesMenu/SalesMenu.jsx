import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Row, Col, Card } from 'react-materialize';
import { entries } from '../App/App';
import './SalesMenu.scss';

// metadata
const month = 'November';
const currency = 'SGD';
const amount = 80586.54;
const cards = [
  {
    cardId: 0,
    title: `Total sales in ${month}?`,
    content: `You have sales worth ${currency} ${amount} in the last 30 days.`,
  },
  {
    cardId: 1,
    title: 'Total sales this week?',
    content: 'You have not sold anything this week or you have not uploaded your invoices.',
  },
  {
    cardId: 2,
    title: 'Card #3',
    content: '#3 Lorem ipsum dolor sit amet',
  },
  {
    cardId: 3,
    title: 'Card #4',
    content: '#4 consectetur adipiscing elit',
  },
];


const Header = ({ title, description }) => (
  <div className="menu-header">
    <Row><Col s={12} className="header-title">{title}</Col></Row>
    <Row><Col s={12} className="header-description">{description}</Col></Row>
  </div>
);
Header.propTypes = {
  title: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
};


// The main UI. The display order of cards are stored in the state.
class SalesMenu extends Component {
  constructor(props) {
    super(props);

    this.state = {
      pinnedCard: -1,
    };
  }

  onPinHandler = (cardId) => {
    const { pinnedCard } = this.state;
    if (cardId === pinnedCard) {
      this.setState({ pinnedCard: -1 });
    } else {
      this.setState({ pinnedCard: cardId });
    }
  }

  isCardPinned = (cardId) => {
    const { pinnedCard } = this.state;
    return cardId === pinnedCard;
  }

  createCard = ({ cardId, title, content }) => (
    <PinCard
      id={`pin-card-${cardId}`}
      key={cardId}
      cardId={cardId}
      title={title}
      content={content}
      isPinned={this.isCardPinned(cardId)}
      onPinHandler={this.onPinHandler}
    />
  );

  render() {
    const { pinnedCard } = this.state;
    return (
      <div className="sales-menu">
        <Header title={entries[0].name} description={entries[0].description} />
        <hr />
        {cards.filter(c => c.cardId === pinnedCard).map(this.createCard)}
        {cards.filter(c => c.cardId !== pinnedCard).map(this.createCard)}
      </div>
    );
  }
}


// A card that supports pinning or unpinning.
// Only one card can be pinned at a time. When one card is pinned, others are automatically
//   unpinned and restored to the default order.
const PinCard = ({
  cardId, title, content, isPinned, onPinHandler,
}) => (
  <Row key={cardId} className="pin-card">
    <Col s={12}>
      <Card
        className="hoverable z-depth-2"
        title={title}
        actions={[
          isPinned
            ? <PinButton cardId={cardId} key={cardId} onPinHandler={onPinHandler} toPin={false} />
            : <PinButton cardId={cardId} key={cardId} onPinHandler={onPinHandler} toPin />,
        ]}
      >
        {content}
      </Card>
    </Col>
  </Row>
);
PinCard.propTypes = {
  cardId: PropTypes.number.isRequired,
  title: PropTypes.string.isRequired,
  content: PropTypes.string.isRequired,
  isPinned: PropTypes.bool.isRequired,
  onPinHandler: PropTypes.func.isRequired,
};


// A flat button that shows either pin or unpin according to the current pinning state.
const PinButton = ({ toPin, cardId, onPinHandler }) => (
  <button
    id={`pin-button-${cardId}`}
    type="button"
    className="btn-flat indigo-text text-darken-2"
    onClick={() => onPinHandler(cardId)}
  >
    {toPin ? 'PIN IT' : 'UNPIN IT'}
  </button>
);
PinButton.propTypes = {
  cardId: PropTypes.number.isRequired,
  toPin: PropTypes.bool.isRequired,
  onPinHandler: PropTypes.func.isRequired,
};

export {
  PinCard,
};

export default SalesMenu;
