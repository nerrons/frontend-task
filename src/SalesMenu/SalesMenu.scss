@import '../index.scss';
@import '../responsive.scss';

.header-title {
  font-size: 3rem;
  color: $primary-color;
  line-height: 2rem;
}

.header-description {
  color: $primary-color;
  line-height: 3rem;
}

.card-title {
  font-weight: 600 !important;
  margin-top: -7px !important;
  font-size: 1rem !important;
}

// The .grey-text and .text-darken-4 are hard coded into react-materialize.
// Piling up specifity to win over the !important rule on .grey-text.
span.card-title.grey-text.text-darken-4 {
  color: $primary-color !important;
}

.card-title + div {
  font-size: 1.2rem !important;
  line-height: 3rem !important; 
}

.card-action {
  color: $primary-color;
  text-align: right;
}

button {
  color: $primary-color;
}