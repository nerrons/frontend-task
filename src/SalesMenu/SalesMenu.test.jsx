import React from 'react';
import ReactDOM from 'react-dom';
import { HashRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import SalesMenu, { PinCard } from './SalesMenu';

Enzyme.configure({ adapter: new Adapter() });

describe('SalesMenu', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<HashRouter><SalesMenu /></HashRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('has a valid snapshot', () => {
    const tree = renderer.create(<HashRouter><SalesMenu /></HashRouter>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders 4 cards', () => {
    const wrapper = shallow(<HashRouter><SalesMenu /></HashRouter>);
    expect(wrapper.find(SalesMenu).dive().find(PinCard).length).toBe(4);
  });

  it('records { pinnedCard: 3 } when Card #4 is pinned', () => {
    const wrapper = mount(<HashRouter><SalesMenu /></HashRouter>);
    wrapper.find('#pin-button-3').simulate('click', { cardId: 3 });
    expect(wrapper.find(SalesMenu).state('pinnedCard')).toBe(3);
  });

  it('shows Card #3 first when Card #3 is pinned', () => {
    const wrapper = mount(<HashRouter><SalesMenu /></HashRouter>);
    wrapper.find('#pin-button-2').simulate('click', { cardId: 2 });
    expect(wrapper.find(SalesMenu).find(PinCard).first().prop('cardId')).toBe(2);
  });

  it('shows Card #3 first when pins #2 then #3', () => {
    const wrapper = mount(<HashRouter><SalesMenu /></HashRouter>);
    wrapper.find('#pin-button-2').simulate('click', { cardId: 2 });
    wrapper.find('#pin-button-3').simulate('click', { cardId: 3 });
    expect(wrapper.find(SalesMenu).find(PinCard).first().prop('cardId')).toBe(3);
  });
});
