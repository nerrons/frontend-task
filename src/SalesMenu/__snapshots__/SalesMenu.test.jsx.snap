// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SalesMenu has a valid snapshot 1`] = `
<div
  className="sales-menu"
>
  <div
    className="menu-header"
  >
    <div
      className="row"
    >
      <div
        className="col s12 header-title"
      >
        Sales
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col s12 header-description"
      >
        Insights about sales
      </div>
    </div>
  </div>
  <hr />
  <div
    className="row pin-card"
  >
    <div
      className="col s12"
    >
      <div
        className="hoverable z-depth-2 card"
      >
        <div
          className="card-content"
        >
          <span
            className="card-title grey-text text-darken-4"
          >
            Total sales in November?
          </span>
          <div>
            You have sales worth SGD 80586.54 in the last 30 days.
          </div>
        </div>
        <div
          className="card-reveal"
        >
          <span
            className="card-title grey-text text-darken-4"
          >
            Total sales in November?
            <i
              className="material-icons right"
            >
              close
            </i>
          </span>
        </div>
        <div
          className="card-action"
        >
          <button
            className="btn-flat indigo-text text-darken-2"
            id="pin-button-0"
            onClick={[Function]}
            type="button"
          >
            PIN IT
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="row pin-card"
  >
    <div
      className="col s12"
    >
      <div
        className="hoverable z-depth-2 card"
      >
        <div
          className="card-content"
        >
          <span
            className="card-title grey-text text-darken-4"
          >
            Total sales this week?
          </span>
          <div>
            You have not sold anything this week or you have not uploaded your invoices.
          </div>
        </div>
        <div
          className="card-reveal"
        >
          <span
            className="card-title grey-text text-darken-4"
          >
            Total sales this week?
            <i
              className="material-icons right"
            >
              close
            </i>
          </span>
        </div>
        <div
          className="card-action"
        >
          <button
            className="btn-flat indigo-text text-darken-2"
            id="pin-button-1"
            onClick={[Function]}
            type="button"
          >
            PIN IT
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="row pin-card"
  >
    <div
      className="col s12"
    >
      <div
        className="hoverable z-depth-2 card"
      >
        <div
          className="card-content"
        >
          <span
            className="card-title grey-text text-darken-4"
          >
            Card #3
          </span>
          <div>
            #3 Lorem ipsum dolor sit amet
          </div>
        </div>
        <div
          className="card-reveal"
        >
          <span
            className="card-title grey-text text-darken-4"
          >
            Card #3
            <i
              className="material-icons right"
            >
              close
            </i>
          </span>
        </div>
        <div
          className="card-action"
        >
          <button
            className="btn-flat indigo-text text-darken-2"
            id="pin-button-2"
            onClick={[Function]}
            type="button"
          >
            PIN IT
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="row pin-card"
  >
    <div
      className="col s12"
    >
      <div
        className="hoverable z-depth-2 card"
      >
        <div
          className="card-content"
        >
          <span
            className="card-title grey-text text-darken-4"
          >
            Card #4
          </span>
          <div>
            #4 consectetur adipiscing elit
          </div>
        </div>
        <div
          className="card-reveal"
        >
          <span
            className="card-title grey-text text-darken-4"
          >
            Card #4
            <i
              className="material-icons right"
            >
              close
            </i>
          </span>
        </div>
        <div
          className="card-action"
        >
          <button
            className="btn-flat indigo-text text-darken-2"
            id="pin-button-3"
            onClick={[Function]}
            type="button"
          >
            PIN IT
          </button>
        </div>
      </div>
    </div>
  </div>
  <a
    aria-current="page"
    className="active"
    href="#/"
    onClick={[Function]}
    style={Object {}}
  >
    Back to Main Menu
  </a>
</div>
`;
